src/vrf.c

```diff
@@ -0,0 +1,236 @@
+        if (ns_rr_class(rr) == ns_c_in && ns_rr_type(rr) == ns_t_mx) {
+            char mxname[NS_MAXDNAME];
+            dn_expand(ns_msg_base(handle), ns_msg_base(handle) + ns_msg_size(handle), ns_rr_rdata(rr) + NS_INT16SZ,
+                      mxname, sizeof mxname);
+            mxs[mx_index++] = strdup(mxname);
+        }
+    }
+
+    return EXIT_SUCCESS;
+}
+
+static int
+check_mx(char *email, struct addrinfo *adrrinfo, Vrf **result)
+{
+    int sock, client_fd;
+    if ((sock = socket(AF_INET, SOCK_STREAM, IPPROTO_IP)) < 0) {
+        printf("Failed to create a socket.");
+        return EXIT_FAILURE;
+    }
+
+    char buffer[SMTP_DATA_LINES_MAX_LENGTH];
+    if ((client_fd = connect(sock, (struct sockaddr *) adrrinfo->ai_addr, sizeof(struct sockaddr))) < 0) {
+        printf("Connection failed.");
+        return EXIT_FAILURE;
+    }
+    read_response(sock, buffer);
+
+    send_command(sock, "EHLO %s\n", CLIENT_MX);
+    read_response(sock, buffer);
+    send_command(sock, "MAIL FROM: <%s>\n", CLIENT_EMAIL);
+    read_response(sock, buffer);
+    send_command(sock, "RCPT TO: <%s>\n", email);
+    read_response(sock, buffer);
+    char status[4];
+    memcpy(status, buffer, 3);
+    status[3] = '\0';
+    long code = strtol(status, NULL, 0);
+    (*result)->result = code == 250;
+
+    send_command(sock, "QUIT\n");
+    close(client_fd);
+
+    return EXIT_SUCCESS;
+}
+
+static void
+email_exists(bool result, bool catch_all, char **verdict)
+{
+    if (catch_all) {
+        *verdict = strdup("may");
+    } else if (result) {
+        *verdict = strdup("does");
+    } else {
+        *verdict = strdup("doesn't");
+    }
+}
+
+void
+free_vrf(Vrf *result)
+{
+    free(result->email);
+    result->email = NULL;
+    free(result->local_part);
+    result->local_part = NULL;
+    free(result->domain);
+    result->domain = NULL;
+    free(result->mx_record);
+    result->mx_record = NULL;
+    free(result->mx_domain);
+    result->mx_domain = NULL;
+    free(result);
+    result = NULL;
+}
+
+void
+print_vrf(FILE *fd, Vrf *result)
+{
+    char *verdict;
+    email_exists(result->result, result->catch_all, &verdict);
+
+    fprintf(fd,
+            "\nVerification summary:\n"
+            "email: %s\n"
+            "local part: %s\n"
+            "domain: %s\n"
+            "mx record: %s\n"
+            "mx domain: %s\n"
+            "result: %s\n"
+            "catch_all: %s\n\n"
+            "It means that this email %s exists!\n\n",
+            result->email,
+            result->local_part,
+            result->domain,
+            result->mx_record,
+            result->mx_domain,
+            result->result ? "true" : "false",
+            result->catch_all ? "true" : "false",
+            verdict
+    );
+
+    free(verdict);
+    verdict = NULL;
+}
+
+int
+verify(Vrf **result)
+{
+    if (extract_local_part_and_domain(result) == EXIT_FAILURE) {
+        printf("Email parts extraction failure.");
+        return EXIT_FAILURE;
```

Comment:
I believe that EXIT_SUCCESS and EXIT_FAILURE are used as a program execution status, not sure if it is used in the right context here in this function? [Source](https://en.cppreference.com/w/c/program/EXIT_status)

Reply:
Agree